src/snowflake/connector/auth/by_plugin.py

```diff
@@ -61,32 +75,113 @@ def reset(self):
         self._current_sleep_time = 1
 
 
-class AuthByPlugin:
+@unique
+class AuthType(Enum):
+    DEFAULT = "SNOWFLAKE"  # default authenticator name
+    EXTERNAL_BROWSER = "EXTERNALBROWSER"
+    KEY_PAIR = "SNOWFLAKE_JWT"
+    OAUTH = "OAUTH"
+    ID_TOKEN = "ID_TOKEN"
+    USR_PWD_MFA = "USERNAME_PASSWORD_MFA"
+    OKTA = "OKTA"
+
+
+class AuthByPlugin(ABC):
     """External Authenticator interface."""
 
     def __init__(self) -> None:
         self._retry_ctx = AuthRetryCtx()
+        self.consent_cache_id_token = False
+        self._timeout = 120
+
+    @property
+    def timeout(self) -> int:
+        return self._timeout
+
+    @timeout.setter
+    def timeout(self, value: Any) -> None:
+        self._timeout = int(value)
 
     @property
-    def assertion_content(self):
+    @abstractmethod
+    def type_(self) -> AuthType:
+        """Return the Snowflake friendly name of auth class."""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def assertion_content(self) -> str:
+        """Return a safe version of the information used to authenticate with Snowflake.
+
+        This is used for logging, useful for printing temporary tokens, but make sure to
+        mask secrets.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def prepare(
+        self,
+        *,
+        conn: SnowflakeConnection,
+        authenticator: str,
+        service_name: str | None,
+        account: str,
+        user: str,
+        password: str | None,
+        **kwargs: Any,
+    ) -> str | None:
+        """Prepare for authentication.
+
+        This function is useful for situations where we need to reach out to a 3rd-party
+        service before authenticating with Snowflake.
+        """
         raise NotImplementedError
 
-    def update_body(self, body):
+    @abstractmethod
+    def update_body(self, body: dict[Any, Any]) -> None:
+        """Update the body of the authentication request."""
         raise NotImplementedError
 
-    def authenticate(self, authenticator, service_name, account, user, password):
+    @abstractmethod
+    def reset_secrets(self) -> None:
+        """Reset secret members."""
         raise NotImplementedError
 
-    def handle_failure(self, ret):
-        """Handles a failure when connecting to Snowflake."""
+    @abstractmethod
+    def reauthenticate(
+        self,
+        *,
+        conn: SnowflakeConnection,
+        **kwargs: Any,
+    ) -> dict[str, Any]:
+        """Re-perform authentication.
+
+        The difference between this and authentication is that secrets will be removed
+        from memory by the time this gets called.
+        """
+        raise NotImplementedError
+
+    def _handle_failure(
```

Comment:
The underscore simply means that nothing outside of `self` should call it, so it's a private method. That's all it means and it can be implemented but it isn't necessary.